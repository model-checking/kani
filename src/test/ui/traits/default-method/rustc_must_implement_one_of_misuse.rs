#![feature(rustc_attrs)]

#[rustc_must_implement_one_of(a, b)]
//~^ Function not found in this trait
//~| Function not found in this trait
trait Tr0 {}

#[rustc_must_implement_one_of(a, b)]
//~^ Function not found in this trait
trait Tr1 {
    fn a() {}
}

#[rustc_must_implement_one_of(a)]
//~^ the `#[rustc_must_implement_one_of]` attribute must be used with at least 2 args
trait Tr2 {
    fn a() {}
}

#[rustc_must_implement_one_of]
//~^ malformed `rustc_must_implement_one_of` attribute input
trait Tr3 {}

#[rustc_must_implement_one_of(A, B)]
trait Tr4 {
    const A: u8 = 1; //~ Not a function

    type B; //~ Not a function
}

#[rustc_must_implement_one_of(a, b)]
trait Tr5 {
    fn a(); //~ This function doesn't have a default implementation

    fn b(); //~ This function doesn't have a default implementation
}

fn main() {}
