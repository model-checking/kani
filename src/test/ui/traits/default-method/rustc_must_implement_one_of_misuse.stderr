error: malformed `rustc_must_implement_one_of` attribute input
  --> $DIR/rustc_must_implement_one_of_misuse.rs:20:1
   |
LL | #[rustc_must_implement_one_of]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[rustc_must_implement_one_of(function1, function2, ...)]`

error: Function not found in this trait
  --> $DIR/rustc_must_implement_one_of_misuse.rs:3:31
   |
LL | #[rustc_must_implement_one_of(a, b)]
   |                               ^

error: Function not found in this trait
  --> $DIR/rustc_must_implement_one_of_misuse.rs:3:34
   |
LL | #[rustc_must_implement_one_of(a, b)]
   |                                  ^

error: Function not found in this trait
  --> $DIR/rustc_must_implement_one_of_misuse.rs:8:34
   |
LL | #[rustc_must_implement_one_of(a, b)]
   |                                  ^

error: the `#[rustc_must_implement_one_of]` attribute must be used with at least 2 args
  --> $DIR/rustc_must_implement_one_of_misuse.rs:14:1
   |
LL | #[rustc_must_implement_one_of(a)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Not a function
  --> $DIR/rustc_must_implement_one_of_misuse.rs:26:5
   |
LL |     const A: u8 = 1;
   |     ^^^^^^^^^^^^^^^^
   |
note: required by this annotation
  --> $DIR/rustc_must_implement_one_of_misuse.rs:24:1
   |
LL | #[rustc_must_implement_one_of(A, B)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: All `#[rustc_must_implement_one_of]` arguments must be associated function names

error: Not a function
  --> $DIR/rustc_must_implement_one_of_misuse.rs:28:5
   |
LL |     type B;
   |     ^^^^^^^
   |
note: required by this annotation
  --> $DIR/rustc_must_implement_one_of_misuse.rs:24:1
   |
LL | #[rustc_must_implement_one_of(A, B)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: All `#[rustc_must_implement_one_of]` arguments must be associated function names

error: This function doesn't have a default implementation
  --> $DIR/rustc_must_implement_one_of_misuse.rs:33:5
   |
LL |     fn a();
   |     ^^^^^^^
   |
note: required by this annotation
  --> $DIR/rustc_must_implement_one_of_misuse.rs:31:1
   |
LL | #[rustc_must_implement_one_of(a, b)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: This function doesn't have a default implementation
  --> $DIR/rustc_must_implement_one_of_misuse.rs:35:5
   |
LL |     fn b();
   |     ^^^^^^^
   |
note: required by this annotation
  --> $DIR/rustc_must_implement_one_of_misuse.rs:31:1
   |
LL | #[rustc_must_implement_one_of(a, b)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 9 previous errors

